lsif-java/src/main/scala/com/sourcegraph/lsif_java/commands/SnapshotLsifCommand.scala

```diff
@@ -174,14 +176,20 @@ object SnapshotLsifCommand {
         .map(m => m.getIdentifier -> m)
         .toMap
     val hovers: Map[String, String] = monikers.map { case (sym, obj) =>
-      val hoverMessage =
-        for {
-          resultSet <- monikerInverse.get(obj.getId).toList
-          hoverEdge <- hoverEdges.get(resultSet).toList
-          hoverResult <- hoverVertexes.get(hoverEdge).toList
-          contents <- hoverResult.getContentsList.asScala
-        } yield contents.getValue.trim
-      sym -> hoverMessage.mkString("\n\n")
+      val hover =
+        (
+          for {
+            resultSet <- monikerInverse.get(obj.getId).toList
+            hoverEdge <- hoverEdges.get(resultSet).toList
+            hoverResult <- hoverVertexes.get(hoverEdge).toList
+            contents <- hoverResult.getContents.getValue.trim
```

Comment:
Here you're iterating through every character of the string only to `mkString` it later. You can yield the value directly instead
```suggestion
            } yield hoverResult.getContents.getValue.trim
```

Reply:
What I would do instead is to replace the enclosing `).mkString(....)` logic with more stuff inside the for comprehension
```scala
        for {
          resultSet <- monikerInverse.get(obj.getId).toList
          hoverEdge <- hoverEdges.get(resultSet).toList
          hoverResult <- hoverVertexes.get(hoverEdge).toList
          contents = hoverResult.getContents.getValue.trim
          codeFence <- contents
            .linesIterator
            .dropWhile(_ != "```java")
            .drop(1)
            .takeWhile(_ != "```")
        } yield codeFence
```

This logic won't crash on an empty string and it only returns an actual result if there is a code fence.

Reply:
Genuinely dont know what all this scala stuff does well enough to be able to change it confidently :sweat_smile:

Reply:
Fair enough. I'm sorry about that, it takes a while to get used to them but once you get the hang of it you can write super concise code and it's difficult to go back. Don't hesitate to ping me directly if you want to update a for comprehension and we can do a walkthrough of how they work.

Reply:
For comprehensions are awesome because you can write like 5 nested for loops (or nested map/flatMap/filter) as a single flat block.